subprojects/core-api/src/main/java/org/gradle/api/attributes/Attribute.java

```diff
@@ -88,6 +109,16 @@ public String getName() {
         return type;
     }
 
+    /**
+     * Determines if this attribute can be included on a published configuration.
+     * @return {@code true} if so; {@code false} otherwise
+     * @since 7.4
+     */
+    @Incubating
+    public boolean isPublishable() {
```

Comment:
I do not like exposing this on the public `Attribute` API.
If that was the only problem, it might be worth it given how this makes for a quite clean implementation.
Especially as this solution then offers an API to plugin authors.

But because of the way published attributes work, there is no difference between having a custom typed or a String typed attribute named `org.gradle.testsuitetype` and so the check done here can easily be worked around.

The alternative could be to hardcode attribute names or even name / value pair into the checking logic. Which could then be only String based - which is also not ideal.

Is the above reflection giving you any alternative idea?

Reply:
I am a bit confused by this PR. The ability to be published is not an inherent quality of an attribute. It's a configuration thing which is orthogonal to the attributes. Therefore it shouldn't be carried by attributes. There is no reason why 2 attributes with the same type and name are considered different if they have a different "publishable" type. IMHO this is not the right model.